Nodejs/Product/Nodejs/TypingsAcquisition.cs

```diff
@@ -0,0 +1,123 @@
+            IEnumerable<IPackage> packages,
+            Redirector redirector) {
+            if (!packages.Any()) {
+                return true;
+            }
+
+            var tsdPath = GetTsdPath(pathToRootNpmDirectory);
+            if (tsdPath == null) {
+                if (redirector != null)
+                    redirector.WriteErrorLine(SR.GetString(SR.TsdNotInstalledError));
+                return false;
+            }
+
+            using (var process = ProcessOutput.Run(
+                tsdPath,
+                TsdInstallArguments(packages),
+                pathToRootProjectDirectory,
+                null,
+                false,
+                redirector,
+                quoteArgs: true)) {
+                if (!process.IsStarted) {
+                    // Process failed to start, and any exception message has
+                    // already been sent through the redirector
+                    if (redirector != null) {
+                        redirector.WriteErrorLine("could not start tsd");
+                    }
+                    return false;
+                }
+                var i = await process;
+                if (i == 0) {
+                    if (redirector != null) {
+                        redirector.WriteLine(SR.GetString(SR.TsdInstallCompleted));
+                    }
+                    return true;
+                } else {
+                    process.Kill();
+                    if (redirector != null) {
+                        redirector.WriteErrorLine(SR.GetString(SR.TsdInstallErrorOccurred));
+                    }
+                    return false;
+                }
+            }
+        }
+
+        private static string GetTsdPath(string pathToRootNpmDirectory) {
+            var path = Path.Combine(pathToRootNpmDirectory, TsdExe);
+            return File.Exists(path) ? path : null;
+        }
+
+        private static string GetPackageTsdName(IPackage package) {
+            return package.Name;
+        }
+
+        private static IEnumerable<string> TsdInstallArguments(IEnumerable<IPackage> packages) {
+            return new[] { "install", }.Concat(packages.Select(GetPackageTsdName)).Concat(new[] { "--save" });
+        }
+
+        private static IEnumerable<IPackage> TypingsToAcquire(string pathToRootProjectDirectory, IEnumerable<IPackage> packages) {
+            var currentTypings = new HashSet<string>(CurrentTypingsPackages(pathToRootProjectDirectory));
+            return packages.Where(package =>
+                !currentTypings.Contains(GetPackageTsdName(package)));
+        }
+
+        private static IEnumerable<string> CurrentTypingsPackages(string pathToRootProjectDirectory) {
+            var typingsDirectoryPath = Path.Combine(pathToRootProjectDirectory, TypingsDirecctoryName);
+            if (Directory.Exists(typingsDirectoryPath)) {
+                foreach (var file in Directory.EnumerateFiles(typingsDirectoryPath, "*.d.ts", SearchOption.AllDirectories)) {
```

Comment:
We need some more robust error handling here... should at least catch IOException.